src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.CoreRT.cs

```diff
@@ -0,0 +1,476 @@
+
+            internal volatile CreateComInterfaceFlagsEx Flags;
+            const ulong ComRefCountMask = 0x000000007fffffffUL;
+            static uint GetComCount(ulong c)
+            {
+                return (uint)(c & ComRefCountMask);
+            }
+
+            public uint AddRef()
+            {
+                return GetComCount((ulong)Interlocked.Increment(ref RefCount));
+            }
+
+            public uint Release()
+            {
+                if (GetComCount((ulong)RefCount) == 0)
+                {
+                    Debug.Fail("Over release of MOW - COM");
+                    return unchecked((uint)-1);
+                }
+
+                return GetComCount((ulong)Interlocked.Decrement(ref RefCount));
+            }
+
+            public unsafe int QueryInterface(in Guid riid, out IntPtr ppvObject)
+            {
+                ppvObject = AsRuntimeDefined(in riid);
+                if (ppvObject == IntPtr.Zero)
+                {
+                    ppvObject = AsUserDefined(in riid);
+                    if (ppvObject == IntPtr.Zero)
+                        return HResults.COR_E_INVALIDCAST;
+                }
+
+                AddRef();
+                return HResults.S_OK;
+            }
+
+            public IntPtr As(in Guid riid)
+            {
+                // Find target interface and return dispatcher or null if not found.
+                IntPtr typeMaybe = AsRuntimeDefined(in riid);
+                if (typeMaybe == IntPtr.Zero)
+                    typeMaybe = AsUserDefined(in riid);
+
+                return typeMaybe;
+            }
+
+            unsafe IntPtr AsRuntimeDefined(in Guid riid)
+            {
+                if ((Flags & CreateComInterfaceFlagsEx.CallerDefinedIUnknown) == CreateComInterfaceFlagsEx.None)
+                {
+                    if (riid == IID_IUnknown)
+                    {
+                        return (IntPtr)(Dispatches + UserDefinedCount);
+                    }
+                }
+
+                return IntPtr.Zero;
+            }
+
+            unsafe IntPtr AsUserDefined(in Guid riid)
+            {
+                for (int i = 0; i < UserDefinedCount; ++i)
+                {
+                    if (UserDefined[i].IID == riid)
+                    {
+                        return (IntPtr)(Dispatches + i);
+                    }
+                }
+
+                return IntPtr.Zero;
+            }
+        }
+
+        internal unsafe class ManagedObjectWrapperHolder
+        {
+            private ManagedObjectWrapper* wrapper;
+            public ManagedObjectWrapperHolder(ManagedObjectWrapper* wrapper)
+            {
+                this.wrapper = wrapper;
+            }
+
+            public unsafe IntPtr ComIp => this.wrapper->As(in ComWrappers.IID_IUnknown);
+
+            ~ManagedObjectWrapperHolder()
+            {
+                // Release GC handle created when MOW was built.
```

Comment:
What happens when the managed object is garbage collected, but the unmanaged IUnknown has non-zero refcount? There does not seem to be anything to handle this situation.

Reply:
@jkotas  I do not account for this case. Should I call `GC.ReRegisterForFinalize(this)` in this case? 
I get additional handle for object and store into Target, My understanding that  until `ManagedObjectWrapper` would be finilized, that object always live since Target hold normal GC handle. Also I have hopes that ManagedObjectWrapper would not be released, since ManagedObjectWrapperHolder has pointer to it. 

What I found I do not do, is that on last call to Release, I should remove object from the _ccwTable and release handle on the `instance`.

Reply:
Also C++ implementation seems to be not accounted for this. Only asserts on DestroySentinel in AddRef/Release presented. I may obviously miss something. My understanding that on Release build anyway this will crash. Isn't it?

Reply:
> What I found I do not do, is that on last call to Release, I should remove object from the _ccwTable and release handle on the `instance`.

Yes. If I am reading the code correctly, the `ManagedObjectWrapperHolder` won't ever get finalized without this.

Reply:
What about other my comment? Do I properly understand chain of ownership, or `ManagedObjectWrapperHolder` does not keep reference to `ManagedObjectWrapper`?

Reply:
Yes, that's sounds right.

Could you please also add a simple tests for the ComWrappers to `src\tests\nativeaot\SmokeTests\PInvoke`? The test can verify that things get released correctly at the right time.